### PR TITLE
Add 'Run script' command to branches pane context menu

### LINKED ISSUE #6307: Branches pane context menu should support "Run script" command
## Feature description
The branches context menu is missing a number of useful commands only available in the commit graph window.  

image

One class of such commands are the "Run Script" options available when right-clicking in the commit graph:

image

The above shows a custom script I've added for interacting with commits.

## Environment

- Git Extensions 3.0.2.5232
- Build 23b6f51905006ccdde8701591df706284b4155dc
- Git 2.19.0.gvfs.1.34.gc7fb556
- Microsoft Windows NT 10.0.18343.0
- .NET Framework 4.8.3745.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by ivangrek**
Fixes #6307

## Test methodology 

- Manually

## Test environment(s) 

- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.23.0.windows.1 (recommended: 2.25.1 or later)
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4200.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
@ivangrek could you please rebase this, and condition the script menu on `IsVisible`, so we can take this?

**Comment by ivangrek** [comment]
> @ivangrek could you please rebase this, and condition the script menu on IsVisible, so we can take this?

Looking forward to #6339 to prevent duplicate work.

File: GitUI/Script/IRunScriptCommand.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace GitUI.Script
+{
+    public interface IRunScriptCommand
```
Comment (mstv) [inline GitUI/Script/IRunScriptCommand.cs:3]: Why not just
```suggestion
    public interface IRunScript
```
?

File: GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs [hunk 2]
```diff
@@ -10,90 +10,104 @@ namespace GitUI.CommandsDialogs
...
 
-            void RemoveOwnScripts()
+            var scripts = ScriptManager.GetScripts();
+            foreach (var script in scripts)
             {
-                hostMenuItem.DropDown.Items.Clear();
+                if (!script.Enabled)
+                {
+                    continue;
+                }
```
Comment (mstv) [inline GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs:41]: Use LINQ `GetScripts().Where(script => script.Enabled)`?

File: GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs [hunk 3]
```diff
@@ -10,90 +10,104 @@ namespace GitUI.CommandsDialogs
...
-                if (lastIndex != contextMenu.Items.Count)
+        private static void RemoveOwnScripts(ContextMenuStrip contextMenu, ToolStripMenuItem hostMenuItem)
+        {
+            hostMenuItem.DropDown.Items.Clear();
+
+            var list = contextMenu.Items.Cast<ToolStripItem>().ToList();
+
+            foreach (var item in list)
+            {
+                if (item.Name.Contains(ScriptNameSuffix))
```
Comment (mstv) [inline GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs:98]: Use LINQ and more precise condition `.Where(item => item.Name.EndsWith(ScriptNameSuffix)`.
Can `.ToList()` be omitted?

File: GitUI/BranchTreePanel/RepoObjectsTree.cs [hunk 4]
```diff
@@ -43,6 +43,7 @@ public partial class RepoObjectsTree : GitModuleControl
         private bool _searchCriteriaChanged;
         private ICheckRefs _refsSource;
         private IScriptHostControl _scriptHost;
+        private IRunScriptCommand _runScriptCommand;
```
Comment (mstv) [inline GitUI/BranchTreePanel/RepoObjectsTree.cs:46]:
```suggestion
        private IRunScript _ScriptRunner;
```
?

File: GitUI/BranchTreePanel/RepoObjectsTree.ContextActions.cs [hunk 5]
```diff
@@ -122,6 +122,15 @@ private void ContextMenuBranchSpecific(ContextMenuStrip contextMenu)
             _localBranchMenuItems.GetInactiveBranchItems().ForEach(t => t.Item.Visible = isNotActiveBranch);
 
             _menuBranchCopyContextMenuItems.ForEach(x => x.Visible = node.Visible);
+
+            if (node.Visible)
+            {
+                contextMenu.AddUserScripts(runScriptToolStripMenuItem, scriptKey => _runScriptCommand.Execute(scriptKey));
```
Comment (mstv) [inline GitUI/BranchTreePanel/RepoObjectsTree.ContextActions.cs:128]:
```suggestion
                contextMenu.AddUserScripts(runScriptToolStripMenuItem, _runScriptCommand.Execute);
```
would reduce noise. Though I have not checked whether it works in C#.

**Review by mstv — COMMENTED** [review]
LGTM, just a few nits / open points regarding naming an C# features.

File: GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs (hunk 3 above)
Comment (ivangrek) [reply GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs:98]: > Can .ToList() be omitted?

Can't iterate and remove item from collection.
We need to create a copy of the collection.

File: GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs [hunk 6]
```diff
@@ -10,90 +10,100 @@ namespace GitUI.CommandsDialogs
...
+                contextMenu.Items.RemoveByKey(item.Name);
             }
+
+            if (contextMenu.Items[contextMenu.Items.Count - 1] is ToolStripSeparator)
+            {
+                contextMenu.Items.RemoveAt(contextMenu.Items.Count - 1);
+            }
+
+            bool showScriptsMenu = hostMenuItem.DropDown.Items.Count > 0;
+            hostMenuItem.Visible = showScriptsMenu;
```
Comment (mstv) [inline GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs:106]: It works without `.ToList()`:
```suggestion
            hostMenuItem.DropDown.Items.Clear();
            hostMenuItem.Visible = false;

            contextMenu.Items.Cast<ToolStripItem>()
                .Where(item => item.Name.EndsWith(ScriptNameSuffix))
                .ForEach(item => contextMenu.Items.RemoveByKey(item.Name));

            if (contextMenu.Items[contextMenu.Items.Count - 1] is ToolStripSeparator)
            {
                contextMenu.Items.RemoveAt(contextMenu.Items.Count - 1);
            }
```

Comment (ivangrek) [reply GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs:106]: image

image

Comment (mstv) [reply GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs:106]: Thank you for the explanation!

Comment (mstv) [reply GitUI/CommandsDialogs/UserScriptContextMenuExtensions.cs:106]: So one final nit from my side:
```suggestion
            hostMenuItem.DropDown.Items.Clear();
            hostMenuItem.Visible = false;

            var list = contextMenu.Items.Cast<ToolStripItem>()
                .Where(x => x.Name.EndsWith(ScriptNameSuffix))
                .ToList();

            foreach (var item in list)
            {
                contextMenu.Items.RemoveByKey(item.Name);
            }

            if (contextMenu.Items[contextMenu.Items.Count - 1] is ToolStripSeparator)
            {
                contextMenu.Items.RemoveAt(contextMenu.Items.Count - 1);
            }
```

**Review by mstv — APPROVED** [review]
Thank you

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
Squash and it is ready to merge. Thank you

**Review by RussKie — APPROVED** [review]

**Pull request merged**